node/podcontroller.go

```diff
@@ -372,21 +388,39 @@ func (pc *PodController) syncHandler(ctx context.Context, key string) error {
 			span.SetStatus(err)
 			return err
 		}
-		// At this point we know the Pod resource doesn't exist, which most probably means it was deleted.
-		// Hence, we must delete it from the provider if it still exists there.
-		if err := pc.deletePod(ctx, namespace, name); err != nil {
-			err := pkgerrors.Wrapf(err, "failed to delete pod %q in the provider", loggablePodNameFromCoordinates(namespace, name))
-			span.SetStatus(err)
+		_, ok := pc.podForDeletion.Load(key)
+		// We've never tried to delete this pod before. The pod was removed completely from API server, so
+		// we treat this as a non-graceful delete.
+		if !ok {
+			pod, err = pc.provider.GetPod(ctx, namespace, name)
+			if errdefs.IsNotFound(err) || (err == nil && pod == nil) {
+				return nil
+			} else if err != nil {
+				err = pkgerrors.Wrapf(err, "failed to fetch pod with key %q from provider", key)
+				span.SetStatus(err)
+				return err
+			}
+			err = pc.provider.DeletePod(ctx, pod)
+			if errdefs.IsNotFound(err) {
+				return nil
+			}
+			if err != nil {
+				err = pkgerrors.Wrapf(err, "failed to delete pod %q in the provider", loggablePodNameFromCoordinates(namespace, name))
+				span.SetStatus(err)
+			}
 			return err
 		}
+		// We tried to delete the pod before. That means we can bail out here, and since this is the last time we are
+		// going to see the pod being updates from Kubernetes, we can remove the pod from the pod deletion map
+		pc.podForDeletion.Delete(key)
```

Comment:
What happens to this in the case we don't get a pod status update to trigger this deletion (common example would be a non-async provider)?

Reply:
You mean, that if a goes into succeeded, or failed without updating any of its container statuses to terminal statuses? If it's in `Running`, then eventually, the fetchPodStatuses will fetch against it and update the pod status.

Reply:
If it goes into suceeded, or failed, we can keep polling the provider. How do providers handle this now?

Reply:
Now deletion is handled synchronously.
What I'm worried about is this not getting cleaned up, and likewise for the API server.

The pod status loop (the one that periodically hits sync providers) clean this out?

Reply:
We could also remove support for sync providers in core and make sure providers are aware they need to trigger a pod status update after delete has occurred.

Reply:
I am in favor for the latter and cut 2.0. We can keep 1.x for bug-fixes.

Reply:
I still think we a GC loop to handle edge cases like missed terminal status notifications or other potential failures.
We can always throw big warning messages in the logs to signal that there may be a bug in the provider for anything we do end up having to clean up.